lib/http/request/writer.rb

```diff
@@ -51,12 +51,16 @@ def connect_through_proxy
       # Adds the headers to the header array for the given request body we are working
       # with
       def add_body_type_headers
-        if @body.is_a?(String) && !@headers[Headers::CONTENT_LENGTH]
-          @request_header << "#{Headers::CONTENT_LENGTH}: #{@body.bytesize}"
-        elsif @body.nil? && !@headers[Headers::CONTENT_LENGTH]
-          @request_header << "#{Headers::CONTENT_LENGTH}: 0"
-        elsif @body.is_a?(Enumerable) && CHUNKED != @headers[Headers::TRANSFER_ENCODING]
-          raise(RequestError, "invalid transfer encoding")
+        unless @headers[Headers::CONTENT_LENGTH]
+          if @body.is_a?(String)
+            @request_header << "#{Headers::CONTENT_LENGTH}: #{@body.bytesize}"
+          elsif @body.respond_to?(:read)
+            @request_header << "#{Headers::CONTENT_LENGTH}: #{@body.size}"
```

Comment:
Yes, when `IO.pipe` is passed in *and* `Content-Length` is not explicitly passed in, then the request will fail with a `NoMethodError` because `IO.pipe` doesn't respond to `#size`. This can be solved by passing `Content-Length` explicitly:

```rb
reader, writer = IO.pipe

writer.write("content")
writer.close

HTTP.headers("Content-Length" => 7).post("http://example.org", body: reader)
```

If the exception isn't raised when `Content-Length` is absent and the IO doesn't respond to `#size`, the user could accidentally send the request without a `Content-Length`, which isn't a valid HTTP request (unless it's a chunked request).

The error is currently just a `NoMethodError`, which probably doesn't communicate well to the user, would it help if I improved the error message? Or are you in disagreement on this behaviour in general?